File: computing/interview/design-patterns/behavioral/mediator/java/MediatorDemo.java
package interview.designpatterns.behavioral.mediator;

import java.util.ArrayList;
import java.util.List;

interface ChatMediator {
    void register(ChatUser user);
    void broadcast(String from, String message);
}

final class ChatRoom implements ChatMediator {
    private final List<ChatUser> users = new ArrayList<>();

    @Override
    public void register(ChatUser user) {
        users.add(user);
    }

    @Override
    public void broadcast(String from, String message) {
        for (ChatUser user : users) {
            if (!user.name().equals(from)) {
                user.receive(from, message);
            }
        }
    }
}

final class ChatUser {
    private final String name;
    private final ChatMediator mediator;

    ChatUser(String name, ChatMediator mediator) {
        this.name = name;
        this.mediator = mediator;
    }

    String name() {
        return name;
    }

    void join() {
        mediator.register(this);
    }

    void send(String message) {
        mediator.broadcast(name, message);
    }

    void receive(String from, String message) {
        System.out.printf("%s received from %s: %s%n", name, from, message);
    }
}

public final class MediatorDemo {
    private MediatorDemo() {}

    public static void main(String[] args) {
        ChatRoom room = new ChatRoom();
        ChatUser alice = new ChatUser("Alice", room);
        ChatUser bob = new ChatUser("Bob", room);
        ChatUser carol = new ChatUser("Carol", room);

        alice.join();
        bob.join();
        carol.join();

        alice.send("Hi everyone!");
        bob.send("Hello Alice!");
    }
}

File: computing/interview/design-patterns/behavioral/mediator/csharp/MediatorDemo.cs
using System;
using System.Collections.Generic;

namespace Interview.DesignPatterns.Behavioral.Mediator
{
    public interface IChatMediator
    {
        void Register(ChatUser user);
        void Broadcast(string from, string message);
    }

    public sealed class ChatRoom : IChatMediator
    {
        private readonly List<ChatUser> _users = new();

        public void Register(ChatUser user)
        {
            _users.Add(user);
        }

        public void Broadcast(string from, string message)
        {
            foreach (var user in _users)
            {
                if (!user.Name.Equals(from, StringComparison.Ordinal))
                {
                    user.Receive(from, message);
                }
            }
        }
    }

    public sealed class ChatUser
    {
        private readonly IChatMediator _mediator;

        public ChatUser(string name, IChatMediator mediator)
        {
            Name = name;
            _mediator = mediator;
        }

        public string Name { get; }

        public void Join()
        {
            _mediator.Register(this);
        }

        public void Send(string message)
        {
            _mediator.Broadcast(Name, message);
        }

        public void Receive(string from, string message)
        {
            Console.WriteLine($"{Name} received from {from}: {message}");
        }
    }

    public static class Demo
    {
        public static void Main()
        {
            var room = new ChatRoom();
            var alice = new ChatUser("Alice", room);
            var bob = new ChatUser("Bob", room);
            var carol = new ChatUser("Carol", room);

            alice.Join();
            bob.Join();
            carol.Join();

            alice.Send("Hi everyone!");
            bob.Send("Hello Alice!");
        }
    }
}
